spring-context/src/main/java/org/springframework/context/PayloadApplicationEvent.java

```diff
@@ -34,22 +36,38 @@
 
 	private final T payload;
 
+	@Nullable
+	private ResolvableType payloadType;
 
 	/**
 	 * Create a new PayloadApplicationEvent.
 	 * @param source the object on which the event initially occurred (never {@code null})
 	 * @param payload the payload object (never {@code null})
 	 */
 	public PayloadApplicationEvent(Object source, T payload) {
+		this(source, payload, null);
+	}
+
+	/**
+	 * Create a new PayloadApplicationEvent.
+	 * @param source the object on which the event initially occurred (never {@code null})
+	 * @param payload the payload object (never {@code null})
+	 * @param payloadType the type object of payload object (can be {@code null})
+	 */
+	public PayloadApplicationEvent(Object source, T payload, @Nullable ResolvableType payloadType) {
 		super(source);
 		Assert.notNull(payload, "Payload must not be null");
 		this.payload = payload;
+		this.payloadType = payloadType;
 	}
 
 
 	@Override
 	public ResolvableType getResolvableType() {
-		return ResolvableType.forClassWithGenerics(getClass(), ResolvableType.forInstance(getPayload()));
+		if(this.payloadType == null){
+			this.payloadType = ResolvableType.forInstance(getPayload());
```

Comment:
Isn't this supposed to build the full nested `PayloadApplicationEvent` type still:
`this.payload = ResolvableType.forClassWithGenerics(getClass(), ResolvableType.forInstance(getPayload()));`